modules/shared/src/main/scala/org/tessellation/schema/currencyMessage.scala

```diff
@@ -0,0 +1,35 @@
+package org.tessellation.schema
+
+import cats.syntax.semigroup._
+
+import org.tessellation.schema.address.Address
+
+import derevo.cats.{eqv, order, show}
+import derevo.circe.magnolia.{decoder, encoder}
+import derevo.derive
+import enumeratum.values.{StringCirceEnum, StringEnum, StringEnumEntry}
+import eu.timepit.refined.auto.{autoInfer, autoRefineV, autoUnwrap}
+import eu.timepit.refined.cats._
+import eu.timepit.refined.types.numeric.NonNegLong
+import io.estatico.newtype.macros.newtype
+
+object currencyMessage {
+  @derive(eqv, show)
+  sealed abstract class MessageType(val value: String) extends StringEnumEntry
+
+  object MessageType extends StringEnum[MessageType] with StringCirceEnum[MessageType] {
+    val values = findValues
+
+    case object Owner extends MessageType("Owner")
+    case object Staking extends MessageType("Staking")
+  }
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class MessageOrdinal(value: NonNegLong) {
+    def next: MessageOrdinal = MessageOrdinal(value |+| 1L)
+  }
+
+  @derive(eqv, show, encoder, decoder)
+  case class CurrencyMessage(messageType: MessageType, address: Address, ordinal: MessageOrdinal)
```

Comment:
should we keep the next ordinal here, or reference to the previous? both should work, but for txs we store the reference 🤔

Reply:
@ryle-goehausen What are your thoughts?

Reply:
We might as well make this consistent with transaction and use reference to previous (ordinal-1), it doesn't particularly matter what the offset is so long as it's changing by 1 each time.

Reply:
1. So the message will contain an optional previous ordinal?
2. If given a previous ordinal then does the new one have to be `previous + 1`? (Seems redundant)
3. Does that mean the message storage only accepts a message if there is none present or if ... what currently stored ordinal matches given previous ordinal?
4. Or if we're to mimic the `create transaction` CLI command, it's taking a file to the previous transaction as input. Is that what we need to do? Receive the previous signed message and construct new one based on that?

Reply:
> So the message will contain an optional previous ordinal?

If you look at the transaction, there is no optional type. The previous ordinal should be always present.

> If given a previous ordinal then does the new one have to be previous + 1? (Seems redundant)

Yes, but you don't need to store the current one. You have the `Next` though (again - check transaction)

> Does that mean the message storage only accepts a message if there is none present or if ... what currently stored ordinal matches given previous ordinal?

Yes, that should be very important (validation).

> Or if we're to mimic the create transaction CLI command, it's taking a file to the previous transaction as input. Is that what we need to do? Receive the previous signed message and construct new one based on that?

It's actually up to you how you want the user to provide it. I see no issue with providing it as CLI parameters

Reply:
@marcinwadon @ryle-goehausen 

So are you saying you want me to mirror `Transaction` and define a `CurrencyMessageReference` and so define `CurrencyMessage` to contain it, like this?
```
case class CurrencyMessage(messageType: MessageType, address: Address, parent: CurrencyMessageReference)
```
or do you want it to follow more closely with how a `Transaction` is defined? e.g.
```
case class CurrencyMessage(
  messageType: MessageType,
  address: Address,
  parent: CurrencyMessageReference,
  salt: CurrencyMessageSalt
) extends Fiber[CurrencyMessageReference, CurrencyMessageData]
  with Encodable[String] { . . .}
```

Reply:
Hey, just to disambiguate, I think you can leave the structure the exact same as it is, and just allow the CLI to pass an integer to it.

```case class CurrencyMessage(messageType: MessageType, address: Address, parentOrdinal: MessageOrdinal)```

I parsed Marcin's comment as the question of should the ordinal be set at N-1 or N (either resulting in the same effect depending how we treat it in the code.) Not whether or not we need a common class or Reference class. So we could could just rename it as parentOrdinal and then define the current ordinal as being parentOrdinal.next() or the other way around. I think that's what he's getting at here.

I don't think we need to change the schema anymore, changing the field name along is fine and providing a current / next method is enough.

Reply:
done